### PR TITLE
feat: detect and ignore massive activations

### LINKED ISSUE #124: Detect and handle massive activations
I've been trying to figure out why the winsorization added in https://github.com/p-e-w/heretic/pull/52 actually helps, and I think I've worked it out: The winsorization indirectly mitigates the influence of ablation on dimensions affected by so-called massive activations.

Massive activations are a phenomenon where specific tokens, sometimes in a position dependent way, trigger massive activations in specific dimensions: https://github.com/locuslab/massive-activations

These activations aren't context sensitive (they have similar values regardless), tend to crop up suddenly in early to middle layers, and tend to persist through several layers before eventually being attenuated.

Due to their magnitude they can bias the mean and end up dominating the ablation direction. More importantly however, ablating them can be catastrophic for the model's performance!

Winsorization mitigates this by setting the magnitude of any massive activations to something more normal. This reduces their impact on the mean direction, but also reduces the impact of ablation on those "special" dimensions.

However, I think a better and more direct way to deal with them is to simply zero out those dimensions in the directions used for ablation (across all layers), so they are left completely untouched.

The more difficult part is actually *detecting* the affected dimensions. There's no real way to infer what tokens might trigger massive activations, so the only way to find them is to examine the hidden state for each token across different prompts.

Detection is heuristic: If a dimension of the hidden state (for a particular token in a particular layer) has `abs_value > 100` and `abs_value > 1000 * abs_layer_median` then it's considered a massive activation. Since a characteristic of massive activations is that they tend to persist for several layers, the detection can be made more robust by checking that large activations are also seen in the next ~4 layers, e.g. with a slightly less strict criterion like `abs_value > 200 * abs_layer_median` to account for attenuation.

I've implemented this locally and it seems to help a lot: For `Gemma-3-12b-it`, dimension 2339 exhibits a lot of massive activations, and simply zeroing it out in the directions leads to similar improvements in refusals / KL divergence as what I've observed with winsorization (though I don't have enough data to confidently say that it's a complete replacement). The full detection identifies a further 3 massive activation dimensions for this model. `Qwen3-4B-Instruct-2507` has two at dimensions 4 and 396, which trigger for a token from its chat template.

Detection can be integrated into `get_residuals` using per-layer hooks, since we already have to set `output_hidden_states=True` anyway. But a small dedicated detection with a prompt containing the most common trigger tokens might also be a good idea (since regular prompts may not contain all of the most common triggers). I've currently implemented a combination of both, and I think it's reasonably clean.

Would you be interested in a PR?

### PR DESCRIPTION
**Pull request by spikymoth**
Closes https://github.com/p-e-w/heretic/issues/124. This is still rough around the edges, containing some hardcoded constants for detection, but I'm not sure how much time I'll have to work on it this weekend. The detection seems pretty effective in my local runs.

There are 5 criteria for detection:
1. The absolute value of the activation must be greater than 100
2. The absolute value of the activation must be greater than 1000x the median of absolute values
3. The absolute deviation from the median of absolute values must be greater than 1000x the MAD
4. It must be part of the cluster of largest outliers and that cluster must contain no more than 3 values
5. The detection must persist for at least 25% of the model's layers

Detection is applied to an initial prompt with common triggers and then to all good and bad prompt residuals.

Finally, we zero out all dimensions with at least 1 detection in both the good and bad residuals. As a result, they won't contribute to the refusal direction and will be left untouched during ablation.

What I'd still like to do before taking this out of draft:
1. Figure out which criteria are most effective at distinguishing massive activations (and maybe finetune the defaults)
2. Remove any redundant criteria and make the rest configurable
3. Make massive activation detection and removal configurable
4. Make the initial detection prompt configurable, maybe?

### CODE REVIEW

File: src/heretic/main.py [hunk 1]
```diff
@@ -418,6 +418,13 @@ def run():
     print("* Obtaining residuals for bad prompts...")
     bad_residuals = model.get_residuals_batched(bad_prompts)
 
+    if model.massive_dims.numel() > 0:
+        # Zero out the dimensions containing massive activations.
+        # These dominate the directions and are very sensitive to ablation.
```
[the remark this answers is not in the document]
Comment (spikymoth) [reply src/heretic/main.py]: Hallucination.

File: src/heretic/model.py [hunk 2]
```diff
@@ -156,6 +157,9 @@ def __init__(self, settings: Settings):
                 f"  * [bold]{component}[/]: [bold]{len(modules)}[/] modules per layer"
             )
 
+        # Initialize the massive activation dimensions.
```
[the remark this answers is not in the document]
Comment (spikymoth) [reply src/heretic/model.py:190]: Hallucination.

File: src/heretic/model.py [hunk 3]
```diff
@@ -618,6 +716,9 @@ def get_residuals(self, prompts: list[Prompt]) -> Tensor:
         # This cast is valid because we passed output_hidden_states=True above.
         hidden_states = cast(tuple[tuple[FloatTensor]], outputs.hidden_states)[0]
 
+        # Identify massive activation dimensions.
```
[the remark this answers is not in the document]
Comment (spikymoth) [reply src/heretic/model.py:825]: Hallucination.

File: src/heretic/model.py [hunk 4]
```diff
@@ -653,6 +754,15 @@ def get_residuals_batched(self, prompts: list[Prompt]) -> Tensor:
 
         return torch.cat(residuals, dim=0)
 
+    def _init_massive_activation_dims(self):
+        prompt = Prompt(
+            system=self.settings.system_prompt,
+            # This prompt contains most common massive activation triggers.
+            user="Summer is warm, and winter\nfrom here doesn't feel cold of all seasons.",
```
[the remark this answers is not in the document]
Comment (spikymoth) [reply src/heretic/model.py:895]: The comment about the comment (sike) is a hallucination. But putting the prompt in a constant or making it configurable is certainly possible.

File: src/heretic/model.py [hunk 5]
```diff
@@ -653,6 +754,15 @@ def get_residuals_batched(self, prompts: list[Prompt]) -> Tensor:
 
         return torch.cat(residuals, dim=0)
 
+    def _init_massive_activation_dims(self):
```
[the remark this answers is not in the document]
Comment (spikymoth) [reply src/heretic/model.py:891]: I wasn't sure about this, as omitting the annotation for void-returning functions seemed somewhat common (e.g. `_apply_lora` and `reset_model`).

File: src/heretic/model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (p-e-w) [reply src/heretic/model.py:895]: Strange that the bot hallucinates such basic issues :(

**Comment by p-e-w** [comment]
Why is the detection logic so complicated and multifaceted? Basically, why isn't the criterion

> The absolute value of the activation must be greater than 1000x the median of absolute values

enough? This seems to perfectly describe what massive activations are, whereas absolute values are pretty suspect since even norms contain learned parameters so they don't really guarantee any particular order of magnitude.

BTW, you might be interested in this post from the creator of MPOA: https://huggingface.co/posts/grimjim/119553185590597

**Comment by spikymoth** [comment]
> Why is the detection logic so complicated and multifaceted? Basically, why isn't the criterion
> 
> > The absolute value of the activation must be greater than 1000x the median of absolute values
> 
> enough? This seems to perfectly describe what massive activations are, whereas absolute values are pretty suspect since even norms contain learned parameters so they don't really guarantee any particular order of magnitude.

The original criteria as proposed by the researchers are points 1 and 2:
1. The absolute value of the activation must be greater than 100
2. The absolute value of the activation must be greater than 1000x the median of absolute values

I think the absolute value is used because activation magnitudes tend to grow from layer to layer, so the absolute value is an easy way to exclude values that aren't astonishingly high in early layers.

I experimented with a bunch of different criteria because these numbers felt kind of arbitrary to me. 1000x the median of absolute values still ignores the scale/dispersion of the values - later layers tend to be more heavy-tailed than earlier layers. I suspect comparing against the MAD (point 3) gives more consistent results.

One of the primary criteria that I found helpful for excluding occasional outliers was consistent detections between layers - but I tried a bunch of different things to look for both false positives and false negatives. I kind of ran out of steam, but I still want to try adding some measurements for different models to see which criteria have the most discriminating power. It's kind of annoying to implement because we want scores for particular tokens from particular prompts in particular layers, then aggregate them into something we can actually draw conclusions from.

Once I have a clearer (data-driven) view of which criteria are most effective, I'll reduce it to 1 or 2 + the cross-layer validation.

> BTW, you might be interested in this post from the creator of MPOA: [huggingface.co/posts/grimjim/119553185590597](https://huggingface.co/posts/grimjim/119553185590597)

Interesting!

**Comment by p-e-w** [comment]
But why do we want to be so conservative? If an activation is 1000x the median, it is always going to dominate computations, *regardless* of whether it also occurs in other layers, of whether it is above a certain absolute magnitude etc. Whenever such an activation pops up, we can assume that the refusal direction will be poor, irrespective of the activation mechanics involved. Those components simply drown out everything else.

**Comment by p-e-w** [comment]
Either way, what are the results with this PR? Have you observed improvements compared to winsorization?

**Comment by p-e-w** [comment]
There might also be a connection with #160, which suggests that more tokens can be "structural" than we currently assume. Perhaps fixing this issue properly, and skipping to the "true" start of the output, can also improve the problem of massive activations.

**Comment by spikymoth** [comment]
> But why do we want to be so conservative? If an activation is 1000x the median, it is always going to dominate computations, _regardless_ of whether it also occurs in other layers, of whether it is above a certain absolute magnitude etc. Whenever such an activation pops up, we can assume that the refusal direction will be poor, irrespective of the activation mechanics involved. Those components simply drown out everything else.

In my experiments, treating too many dimensions as massive actually reduced performance. I think sometimes large activations are just important, and ignoring the dimension containing them just weakens the effectiveness of ablation.

> Either way, what are the results with this PR? Have you observed improvements compared to winsorization?

I've observed *good* performance, but it's very difficult with my hardware to gather a large comparative dataset. I'll try to get you a more concrete answer.

> There might also be a connection with #160, which suggests that more tokens can be "structural" than we currently assume. Perhaps fixing this issue properly, and skipping to the "true" start of the output, can also improve the problem of massive activations.

Yes, I think this is also an important problem - we want to do our measurements on the logits that best identify the difference between acceptance and refusal. But I think moving the measurement position is orthogonal to detecting and ignoring dimensions in the hidden state that exhibit massive activations.

Massive activations are very token-specific and basically don't show up at the end of prompt / start of response position (with the exception of that 1 dimension in Gemma 3, which is just massive all the time), so moving the measurement position won't do anything to reduce the impact of ablating those dimensions on model performance.

IIRC dimensions that exhibit massive activations do generally tend to have larger norms in the weight matrix (though large norms don't unambiguously identify them), so I think they do tend to have a larger than normal impact on the refusal direction. So winsorizing or ignoring them might also help "purify" the direction in that sense too.

**Comment by p-e-w** [comment]
> IIRC dimensions that exhibit massive activations do generally tend to have larger norms in the weight matrix (though large norms don't unambiguously identify them), so I think they do tend to have a larger than normal impact on the refusal direction. So winsorizing or ignoring them might also help "purify" the direction in that sense too.

Perhaps it could also make sense to apply this pruning to magnitude preservation somehow.

As always, the main problem is that we can't clearly tell whether a change is an improvement or not. This is a huge issue that we need to address comprehensively and scientifically. Many potential enhancements depend on us being able to tell the difference between a genuine improvement and noise.

**Comment by p-e-w** [comment]
@spikymoth 

The TurboQuant paper uses random rotations to deal with massive activations insofar as they affect quantization (see [my Reddit post explaining how this works](https://www.reddit.com/r/LocalLLaMA/comments/1s62g5v/a_simple_explanation_of_the_key_idea_behind/)).

While this doesn't eliminate the vector alignment problem, it could make our computations a lot more numerically stable, and perhaps have other benefits as well.

**Comment by spikymoth** [comment]
Funnily enough, I did actually play with rotation matrices a while ago to see if it would improve the direction estimates (I made a robust estimator that worked by applying a rotation, taking the medians and undoing the rotation, then repeating that with several other random rotations and averaging the results). It worked but didn't really outperform other forms of robust estimation.

However, I didn't attempt to apply it further along the chain - e.g. removing a rotated direction from a rotated weight matrix. I'm not sure how well it would work, but it could be interesting as earlier layers tend to be light-tailed while later layers tend to be heavy-tailed, and orthogonal rotations give them all normal tails (which might make the direction apply more universally, maybe?).

I'm not sure how much it would really impact massive activations, though. Hoping to get back to this PR soon.

**Comment by spikymoth** [comment]
Also mostly a rebase, but I did change the logic to keep the massive activation dims intact for determining the refusal direction. Instead we now zero out those dimensions of LoRA component B, to avoid changing them in the ablation.

I haven't done a before/after comparison, but KL divergence and refusal reduction numbers still look fine at first glance. I'll do more with this PR when https://github.com/p-e-w/heretic/pull/138 is further along.

**Comment by p-e-w** [comment]
> Also mostly a rebase, but I did change the logic to keep the massive activation dims intact for determining the refusal direction. Instead we now zero out those dimensions of LoRA component B, to avoid changing them in the ablation.

That's an interesting idea. I'm still trying to develop a theoretical understanding of what is actually going on here, and what the correct approach really is. It's tempting to just discard large components, but of course that's ultimately black magic and only "works" in the sense that we can hope that the distortion we introduce does less damage than the massive activations themselves.

One approach I had in mind is to apply an affine transformation to each residual vector that shifts the origin to the mean of all residual vectors. This makes massive components disappear. All direction computations can then be performed on those transformed vectors. The problem is that such an affine transformation cannot be compensated for by subtracting from the weight matrix, as we do now, so we would have to undo it before calculating the outer product, which pretty much destroys the benefits apart from numerical stability. However, it might be possible to apply the inverse transform *through the bias vector of the module* instead. Unfortunately, I believe that many models don't have bias vectors for projections at all.
